### PR TITLE
CHERI-RISC-V: Add CLoadTags instruction

### PR DESCRIPTION
**Pull request by nwf**

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVInstrInfoXCheri.td [hunk 1]
```diff
@@ -585,6 +585,10 @@ defm SC_C       : AMO_C_rr_aq_rl<"128", 0b00011, 0b100, "sc.c", GPR>;
 defm AMOSWAP_C  : AMO_C_rr_aq_rl<"128", 0b00001, 0b100, "amoswap.c", GPCR>;
 }
 
+let Predicates = [HasCheri], hasSideEffects = 0, mayLoad = 1, mayStore = 0 in
+def CLoadTags : RVInstCheriSrcDst<0x7f, 0x12, 0, OPC_CHERI, (outs GPR:$rd),
+                              (ins GPCR:$cs1), "cloadtags", "$rd, $cs1">;
```
Comment (jrtc27) [inline llvm/lib/Target/RISCV/RISCVInstrInfoXCheri.td:590]: Belongs in a new "Tag-Memory Access Instructions" section just after "Adjusting to Compressed Capability Precision Instructions" (grouped based on the ISAv8 quick reference)

File: llvm/lib/Target/RISCV/RISCVInstrInfoXCheri.td [hunk 2]
```diff
@@ -1122,6 +1124,8 @@ defm : CheriExplicitLdPat<zextloadi32, LWU_CAP, GPCR>;
 defm : CheriExplicitLdPat<load, LC_CAP_128, GPCR>;
 }
 
+def : PatGpcr<int_cheri_cap_load_tags, CLoadTags>;
```
Comment (jrtc27) [inline llvm/lib/Target/RISCV/RISCVInstrInfoXCheri.td:1127]: Ditto

File: llvm/lib/Target/RISCV/RISCVInstrInfoXCheri.td [hunk 3]
```diff
@@ -585,6 +585,10 @@ defm SC_C       : AMO_C_rr_aq_rl<"128", 0b00011, 0b100, "sc.c", GPR>;
 defm AMOSWAP_C  : AMO_C_rr_aq_rl<"128", 0b00001, 0b100, "amoswap.c", GPCR>;
 }
 
+let Predicates = [HasCheri], hasSideEffects = 0, mayLoad = 1, mayStore = 0 in
+def CLoadTags : RVInstCheriSrcDst<0x7f, 0x12, 0, OPC_CHERI, (outs GPR:$rd),
```
Comment (jrtc27) [inline llvm/lib/Target/RISCV/RISCVInstrInfoXCheri.td:589]: I think Cheri_r would work (with `mayLoad = 1`)?

File: llvm/lib/Target/RISCV/RISCVInstrInfoXCheri.td (hunk 1 above)
Comment (jrtc27) [inline llvm/lib/Target/RISCV/RISCVInstrInfoXCheri.td:590]: GPCRMemAtomic would be nicer given it's a load

**Comment by jrtc27** [comment]
TableGen looks good now, but please add a test to llvm/test/MC/RISCV/cheri/rv32xcheri-valid.s (also kept in the same order)

**Comment by jrtc27** [comment]
@arichardson What do you want doing for PRs? Wait for the upstream merge, target upstream-llvm-merge or just merge to dev?

**Comment by nwf** [comment]
LLVM and I are disagreeing about the encoding of `cloadtags` :(

**Comment by arichardson** [comment]
> @arichardson What do you want doing for PRs? Wait for the upstream merge, target upstream-llvm-merge or just merge to dev?

I'd prefer if we could re-target changes right now to the upstream-llvm-merge branch to avoid having to change any tests when merging dev to upstream-llvm-merge (quite a lot has changed in the past few months). This probably doesn't affect this patch, but at targeting upstream-llvm-merge it will avoid another merge commit in the history.

File: llvm/lib/Target/RISCV/RISCVInstrInfoXCheri.td [hunk 4]
```diff
@@ -80,9 +80,9 @@ include "RISCVInstrFormatsXCheri.td"
 
 let hasSideEffects = 0, mayLoad = 0, mayStore = 0 in
 class Cheri_r<bits<5> funct5, string opcodestr, RegisterClass rdClass=GPR,
-              RegisterClass rs1Class=GPCR>
+              DAGOperand rs1Operand=GPCR>
```
Comment (arichardson) [inline llvm/lib/Target/RISCV/RISCVInstrInfoXCheri.td:83]:
```suggestion
              RegisterOperand rs1Operand=GPCR>
```

Does this work?

**Comment by arichardson** [comment]
Additionally, tablegen got a lot stricter with type mismatches in the upstream-llvm-merge branch so we should ideally target that.

File: llvm/lib/Target/RISCV/RISCVInstrInfoXCheri.td (hunk 4 above)
Comment (nwf) [reply llvm/lib/Target/RISCV/RISCVInstrInfoXCheri.td:83]: No. `RegisterClass` is not a subtype of `RegisterOperand`; instead, both of those are sibling subtypes of `DAGOperand`.

https://github.com/llvm/llvm-project/blob/538254e8e0e09e89776f21bd39c23be1f5868fa1/llvm/include/llvm/Target/Target.td#L217-L219

https://github.com/llvm/llvm-project/blob/538254e8e0e09e89776f21bd39c23be1f5868fa1/llvm/include/llvm/Target/Target.td#L847-L848

Comment (arichardson) [reply llvm/lib/Target/RISCV/RISCVInstrInfoXCheri.td:83]: Thanks for the links, I wasn't sure.

**Comment by nwf** [comment]
I'm content to wait to merge until `upstream-llvm-merge` is merged into `dev`; I can just run this locally for a while.

**Comment by arichardson** [comment]
> I'm content to wait to merge until `upstream-llvm-merge` is merged into `dev`; I can just run this locally for a while.

I think all that is missing now is CheriBSD Clang 12 fixes. So hopefully soon.

**Review by arichardson — APPROVED** [review]
Encoding `0xff2100db` seems to match the ISA spec, so LGTM if the tests pass.

**Review by jrtc27 — APPROVED** [review]

**Pull request merged**